### PR TITLE
Process Data Model

### PR DESCRIPTION
**Pull request by aamir-asaram**
**In this PR, I have implemented the following:**
- Created models for blog components:
  - User
  - Post
  - Comment
  - Like
- Implemented custom methods for models

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: 'author_id', class_name: 'Post'
+  has_many :comments
+  has_many :likes
```
Comment (cynthiainga) [inline app/models/user.rb]: - [ ] Good job so far but correctly according to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png) the `comments` and `likes` table has foreign key `author_id`, also for `posts` you don't need to add the `class_name: 'Post` 🙃 🙃

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  has_many :comments
+  has_many :likes
+
+  def recent_comments
+    comments.order(created_at: :desc).limit(5)
+  end
+
+  def increment_posts_counter
+    author.increment!(:posts_counter)
+  end
```
Comment (cynthiainga) [inline app/models/post.rb:16]: - [OPTIONAL] Good job on implementing the `increment_posts_counter`, but I would suggest using one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#increment_posts_counter` method every time a new post has been saved to database, then make the `increment_posts_counter` method private 🙃 
Like 👇🏾 

```ruby
after_save :increment_posts_counter

  private

  def increment_posts_counter
    author.increment!(:posts_counter)
  end
```
- You can do the same for `increment_comments_counter` and `increment_likes_counter`

**Review by cynthiainga — CHANGES_REQUESTED** [review]
Hi @aamir-asaram, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights

- [ ] User model has a method that returns 3 most recent posts ✔️ 
- [ ] Post model has a method that updates the posts counter for a user ✔️ 
- [ ] Post model has a method that returns the 5 most recent comments ✔️ 
- [ ] Comment model has a method that updates the comments counter for a post ✔️ 
- [ ] Like model has a method that updates the likes counter for a post ✔️ 

## Required Changes ♻️

- [ ] Please kindly check the comments under the review 👇🏾 👇🏾 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by aamir-asaram** [comment]
Thank you for the helpful feedback @cynthiainga! I have implemented all your suggestions :smile:

**Review by gemmen29 — APPROVED** [review]
Hi @aamir-asaram, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 



## Optional suggestions

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**